backend/curation/api/v1/curation/datasets/actions.py

```diff
@@ -0,0 +1,33 @@
+from flask import jsonify, make_response
+
+from backend.curation.api.v1.curation.collections.common import reshape_dataset_for_curation_api
+from backend.layers.common.entities import CollectionLinkType
+from backend.portal.api.providers import get_business_logic
+
+
+def get():
+    """
+    Datasets index endpoint to retrieve full metadata for all public Datasets.
+    """
+    collections_with_datasets = get_business_logic().get_all_published_collections_with_datasets()
+
+    all_datasets_with_collection_name_and_doi = []
+
+    for collection in collections_with_datasets:
+        assert collection.datasets  # Any public Collection should have at least one Dataset
+        for dataset in collection.datasets:
+            dataset_response_obj = reshape_dataset_for_curation_api(dataset, is_published=True, use_canonical_id=True)
+            collection_doi = None
+            for link in collection.metadata.links:
+                if link.type == CollectionLinkType.DOI.name:
+                    collection_doi = link.uri
+            dataset_response_obj.update(
```

Comment:
We should only need to build this object once per collection and add it to each dataset object, rather than rebuild every dataset.